// mobile-locator/dist/api/google-geolocation.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = require('lodash/has');
var Base = require('./base');

var API = 'https://www.googleapis.com/geolocation/v1/geolocate';

/* eslint-disable class-methods-use-this */

var GoogleGeolocation = function (_Base) {
  _inherits(GoogleGeolocation, _Base);

  function GoogleGeolocation(options) {
    _classCallCheck(this, GoogleGeolocation);

    var _this = _possibleConstructorReturn(this, (GoogleGeolocation.__proto__ || Object.getPrototypeOf(GoogleGeolocation)).call(this, options));

    if (options) {
      _this.key = options.key;
    }
    return _this;
  }

  _createClass(GoogleGeolocation, [{
    key: 'getRequestSettings',
    value: function getRequestSettings(cell) {
      return {
        method: 'POST',
        uri: API,
        qs: {
          key: this.key
        },
        json: {
          considerIp: false,
          cellTowers: [{
            cellId: cell.cid,
            locationAreaCode: cell.lac,
            mobileCountryCode: cell.mcc,
            mobileNetworkCode: cell.mnc
          }]
        }
      };
    }
  }, {
    key: 'validate',
    value: function validate(body) {
      return !has(body, 'error');
    }
  }, {
    key: 'parseLocation',
    value: function parseLocation(body) {
      return {
        longitude: body.location.lng,
        latitude: body.location.lat,
        accuracy: body.accuracy
      };
    }
  }, {
    key: 'parseError',
    value: function parseError(body) {
      return has(body, 'error') ? body.error.errors[0].reason : body;
    }
  }]);

  return GoogleGeolocation;
}(Base);

module.exports = GoogleGeolocation;
