// mobile-locator/dist/api/gpsspg.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = require('lodash/has');
var Base = require('./base');

var API = 'http://api.gpsspg.com/bs/';

/* eslint-disable class-methods-use-this */

var GPSspg = function (_Base) {
  _inherits(GPSspg, _Base);

  function GPSspg(options) {
    _classCallCheck(this, GPSspg);

    var _this = _possibleConstructorReturn(this, (GPSspg.__proto__ || Object.getPrototypeOf(GPSspg)).call(this, options));

    if (options) {
      _this.key = options.key;
      _this.oid = options.oid;
      _this.system = options.system;
    }
    return _this;
  }

  _createClass(GPSspg, [{
    key: 'getRequestSettings',
    value: function getRequestSettings(cell) {
      return {
        uri: API,
        qs: {
          oid: this.oid,
          type: '',
          mcc: cell.mcc,
          mnc: cell.mnc,
          a: cell.lac,
          b: cell.cid,
          //  0:WGS84; 1:Google Maps; 2:Baidu Maps; 3:QQ Maps; 4:MapBar
          to: this.system || 0,
          key: this.key,
          output: 'json'
        },
        json: true
      };
    }
  }, {
    key: 'validate',
    value: function validate(body) {
      return !has(body, 'error') && body.status === '200';
    }
  }, {
    key: 'parseLocation',
    value: function parseLocation(body) {
      return {
        longitude: body.result.lng,
        latitude: body.result.lat,
        accuracy: body.result.radius,
        address: body.result.address
      };
    }
  }, {
    key: 'parseError',
    value: function parseError(body) {
      return has(body, 'status') ? body.status : body;
    }
  }]);

  return GPSspg;
}(Base);

module.exports = GPSspg;
