// mobile-locator/dist/api/mylnikov.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = require('lodash/has');
var Base = require('./base');

var API = 'https://api.mylnikov.org/geolocation/cell';
var API_VERSION = '1.1';

/* eslint-disable class-methods-use-this */

var Mylnikov = function (_Base) {
  _inherits(Mylnikov, _Base);

  function Mylnikov(options) {
    _classCallCheck(this, Mylnikov);

    var _this = _possibleConstructorReturn(this, (Mylnikov.__proto__ || Object.getPrototypeOf(Mylnikov)).call(this, options));

    if (options) {
      if (options.data) {
        _this.data = options.data;
      }
    }
    return _this;
  }

  _createClass(Mylnikov, [{
    key: 'getRequestSettings',
    value: function getRequestSettings(cell) {
      return {
        uri: API,
        json: true,
        qs: {
          v: API_VERSION,
          data: this.data,
          mcc: cell.mcc,
          mnc: cell.mnc,
          lac: cell.lac,
          cellid: cell.cid
        }
      };
    }
  }, {
    key: 'validate',
    value: function validate(body) {
      return body.result === 200;
    }
  }, {
    key: 'parseLocation',
    value: function parseLocation(body) {
      return {
        longitude: body.data.lon,
        latitude: body.data.lat,
        accuracy: body.data.range
      };
    }
  }, {
    key: 'parseError',
    value: function parseError(body) {
      return has(body, 'desc') ? body.desc : body;
    }
  }]);

  return Mylnikov;
}(Base);

module.exports = Mylnikov;
