// mobile-locator/dist/api/opencellid.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = require('lodash/has');
var Base = require('./base');

var API = 'http://opencellid.org/cell/get';

/* eslint-disable class-methods-use-this */

var OpenCellID = function (_Base) {
  _inherits(OpenCellID, _Base);

  function OpenCellID(options) {
    _classCallCheck(this, OpenCellID);

    var _this = _possibleConstructorReturn(this, (OpenCellID.__proto__ || Object.getPrototypeOf(OpenCellID)).call(this, options));

    if (options) {
      _this.key = options.key;
    }
    return _this;
  }

  _createClass(OpenCellID, [{
    key: 'getRequestSettings',
    value: function getRequestSettings(cell) {
      return {
        uri: API,
        json: true,
        qs: {
          key: this.key,
          mcc: cell.mcc,
          mnc: cell.mnc,
          lac: cell.lac,
          cellid: cell.cid,
          format: 'json'
        }
      };
    }
  }, {
    key: 'validate',
    value: function validate(body) {
      return !has(body, 'error');
    }
  }, {
    key: 'parseLocation',
    value: function parseLocation(body) {
      return {
        longitude: body.lon,
        latitude: body.lat,
        accuracy: body.range
      };
    }
  }, {
    key: 'parseError',
    value: function parseError(body) {
      return has(body, 'error') ? body.error : body;
    }
  }]);

  return OpenCellID;
}(Base);

module.exports = OpenCellID;
