// mobile-locator/dist/api/unwiredlabs.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = require('lodash/has');
var Base = require('./base');

var API = 'https://unwiredlabs.com/v2/process.php';

/*
 * US East (Northern Virginia): https://us1.unwiredlabs.com/v2/process.php
 * US West (San Francisco): https://us2.unwiredlabs.com/v2/process.php
 * Europe (Ireland): https://eu1.unwiredlabs.com/v2/process.php
 * Asia Pacific (Singapore): https://ap1.unwiredlabs.com/v2/process.php
 */

/* eslint-disable class-methods-use-this */

var UnwiredLabs = function (_Base) {
  _inherits(UnwiredLabs, _Base);

  function UnwiredLabs(options) {
    _classCallCheck(this, UnwiredLabs);

    var _this = _possibleConstructorReturn(this, (UnwiredLabs.__proto__ || Object.getPrototypeOf(UnwiredLabs)).call(this, options));

    if (options) {
      _this.token = options.token;
    }
    return _this;
  }

  _createClass(UnwiredLabs, [{
    key: 'getRequestSettings',
    value: function getRequestSettings(cell) {
      return {
        method: 'POST',
        uri: API,
        json: {
          token: this.token,
          radio: 'gsm',
          mcc: cell.mcc,
          mnc: cell.mnc,
          cells: [{
            lac: cell.lac,
            cid: cell.cid
          }],
          address: 1
        }
      };
    }
  }, {
    key: 'validate',
    value: function validate(body) {
      return body.status === 'ok';
    }
  }, {
    key: 'parseLocation',
    value: function parseLocation(body) {
      return {
        longitude: parseFloat(body.lon),
        latitude: parseFloat(body.lat),
        accuracy: parseInt(body.accuracy, 10),
        address: body.address
      };
    }
  }, {
    key: 'parseError',
    value: function parseError(body) {
      return has(body, 'status') && body.status === 'error' ? body.message : body;
    }
  }]);

  return UnwiredLabs;
}(Base);

module.exports = UnwiredLabs;
