// mobile-locator/dist/api/yandex.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var has = require('lodash/has');
var Base = require('./base');

var API = 'http://api.lbs.yandex.net/geolocation';

/* eslint-disable class-methods-use-this */

var Yandex = function (_Base) {
  _inherits(Yandex, _Base);

  function Yandex(options) {
    _classCallCheck(this, Yandex);

    var _this = _possibleConstructorReturn(this, (Yandex.__proto__ || Object.getPrototypeOf(Yandex)).call(this, options));

    if (options) {
      _this.key = options.key;
    }
    return _this;
  }

  _createClass(Yandex, [{
    key: 'getRequestSettings',
    value: function getRequestSettings(cell) {
      var json = {
        common: {
          version: '1.0',
          api_key: this.key
        },
        gsm_cells: [{
          countrycode: cell.mcc,
          operatorid: cell.mnc,
          lac: cell.lac,
          cellid: cell.cid
        }]
      };
      return {
        method: 'POST',
        uri: API,
        body: 'json=' + JSON.stringify(json)
      };
    }
  }, {
    key: 'preprocessBody',
    value: function preprocessBody(body) {
      return JSON.parse(body);
    }
  }, {
    key: 'validate',
    value: function validate(body) {
      try {
        return !has(body, 'error') && body.position.type !== 'ip';
      } catch (e) {
        return false;
      }
    }
  }, {
    key: 'parseLocation',
    value: function parseLocation(body) {
      return {
        longitude: body.position.longitude,
        latitude: body.position.latitude,
        accuracy: body.position.precision
      };
    }
  }, {
    key: 'parseError',
    value: function parseError(body) {
      if (has(body, 'error')) {
        return '[' + body.error.code + '] ' + body.error.text;
      }

      if (body.position.type === 'ip') {
        return 'Cell not found.';
      }

      return body;
    }
  }]);

  return Yandex;
}(Base);

module.exports = Yandex;
